src/context.rs

```diff
@@ -0,0 +1,336 @@
+/// in time and execute the logic in the associated `FnOnce(&mut Context)`
+/// closure. Modules can add plans to this queue through the `Context`.
+///
+/// The simulation also has an immediate callback queue to allow for the
+/// accumulation of side effects of mutations by the current controlling module.
+///
+pub struct Context {
+    plan_queue: Queue<Box<Callback>>,
+    callback_queue: VecDeque<Box<Callback>>,
+    data_plugins: HashMap<TypeId, Box<dyn Any>>,
+    current_time: f64,
+}
+
+impl Context {
+    /// Create a new empty `Context`
+    #[must_use]
+    pub fn new() -> Context {
+        Context {
+            plan_queue: Queue::new(),
+            callback_queue: VecDeque::new(),
+            data_plugins: HashMap::new(),
+            current_time: 0.0,
+        }
+    }
+
+    /// Add a plan to the future event list at the specified time
+    ///
+    /// Returns an `Id` for the newly-added plan that can be used to cancel it
+    /// if needed.
+    /// # Panics
+    ///
+    /// Panics if time is in the past, infinite, or NaN.
+    pub fn add_plan(&mut self, time: f64, callback: impl FnOnce(&mut Context) + 'static) -> Id {
+        assert!(!time.is_nan() && !time.is_infinite() && time >= self.current_time);
+        self.plan_queue.add_plan(time, Box::new(callback))
+    }
+
+    /// Cancel a plan that has been added to the queue
+    ///
+    /// # Panics
+    ///
+    /// This function panics if you cancel a plan which has already been
+    /// cancelled or executed.
+    pub fn cancel_plan(&mut self, id: &Id) {
+        self.plan_queue.cancel_plan(id);
+    }
+
+    /// Add a `Callback` to the queue to be executed before the next plan
+    pub fn queue_callback(&mut self, callback: impl FnOnce(&mut Context) + 'static) {
+        self.callback_queue.push_back(Box::new(callback));
+    }
+
+    /// Retrieve a mutable reference to the data container associated with a
+    /// `DataPlugin`
+    ///
+    /// If the data container has not been already added to the `Context` then
+    /// this function will use the `DataPlugin::create_data_container` method
+    /// to construct a new data container and store it in the `Context`.
+    ///
+    /// Returns a mutable reference to the data container
+    #[must_use]
+    #[allow(clippy::missing_panics_doc)]
+    pub fn get_data_container_mut<T: DataPlugin>(&mut self) -> &mut T::DataContainer {
+        self.data_plugins
+            .entry(TypeId::of::<T>())
+            .or_insert_with(|| Box::new(T::create_data_container()))
+            .downcast_mut::<T::DataContainer>()
+            .unwrap() // Will never panic as data container has the matching type
+    }
+
+    /// Retrieve a reference to the data container associated with a
+    /// `DataPlugin`
+    ///
+    /// Returns a reference to the data container if it exists or else `None`
+    #[must_use]
+    pub fn get_data_container<T: DataPlugin>(&self) -> Option<&T::DataContainer> {
```

Comment:
The idea is that we want this method to return a reference to the data container in the current `Context` if it exists. If it doesn't there isn't a way to return a reference as nothing would own the data container. Circumventing this would entail interior mutability which I think for now we want to avoid.